crux/src/Crux/Goal.hs

```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
           -- NOTE: We don't currently provide a method for capturing the output
           -- sent to offline solvers.  We would probably want a file per goal.
           let logData = WS.defaultLogData
-          mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
-            case satRes of
-              Unsat _ -> do
-                -- NOTE: We don't have an easy way to get an unsat core here
-                -- because we don't have a solver connection.
-                let core = fmap Left (F.toList (mconcat assumptionsInScope))
-                end
-                modifyIORef' goalNum (updateProcessedGoals p (Proved core))
-                return (Prove (p, Proved core))
-              Sat (evalFn, _) -> do
-                let model = ctx ^. cruciblePersonality . personalityModel
-                vals <- evalModel evalFn model
-                explain <- explainFailure evalFn p
-                end
-                let gt = NotProved explain (Just (ModelView vals))
-                modifyIORef' goalNum (updateProcessedGoals p gt)
-                when (failfast && not (isResourceExhausted p)) $
-                  sayOK "Crux" "Counterexample found, skipping remaining goals"
-                return (Prove (p, gt))
-              Unknown -> do
-                end
-                modifyIORef' goalNum (updateProcessedGoals p (NotProved mempty Nothing))
-                return (Prove (p, NotProved mempty Nothing))
-          case mres of
-            Just res -> return res
-            Nothing -> return (Prove (p, NotProved mempty Nothing))
 
+          let runOneSolver adapter = do
+                mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
+                  case satRes of
+                    Unsat _ -> do
+                      -- NOTE: We don't have an easy way to get an unsat core here
+                      -- because we don't have a solver connection.
+                      let core = fmap Left (F.toList (mconcat assumptionsInScope))
+                      return (Proved core)
+                    Sat (evalFn, _) -> do
+                      let model = ctx ^. cruciblePersonality . personalityModel
+                      vals <- evalModel evalFn model
+                      explain <- explainFailure evalFn p
+                      return (NotProved explain (Just (ModelView vals)))
+                    Unknown -> do
+                      return (NotProved mempty Nothing)
+                case mres of
+                  Just res -> return res
+                  Nothing -> return (NotProved mempty Nothing)
```

Comment:
The nothing case corresponds to timeout, right? Probably worth mentioning that in the explanation.

Reply:
Do you happen to know if the timeout corresponds to wall clock or actual CPU time? That is, is a per-thread timeout the same behavior as a a timeout on the `waitAnyCatch`? Not terribly important, but I think `waitAnyCatch` ends up catching the `Timeout` exception, which means that if `withTimeout` returns, it only returns a `Just` value.

Reply:
I think it's wall-clock time, but I'm not sure.  Given the way this is written, I'd expect the `withTimeout` combinator to handle the `Timeout` exception before it bubbles out, but I don't know offhand where that combinator comes from or how it works, exactly.